File: frontend/rdp-security.js
/**
 * RDP Security Policy Module
 * Provides connection security enforcement with immutable policy configuration
 */

/**
 * Deep freeze an object to prevent tampering
 * @param {any} obj - Object to freeze
 * @returns {any} Frozen object
 */
export function deepFreeze(obj) {
    // Handle primitives and null
    if (obj === null || typeof obj !== 'object') return obj;

    // Freeze arrays first: elements may need freezing too
    if (Array.isArray(obj)) {
        for (const el of obj) deepFreeze(el);
        return Object.freeze(obj);
    }

    // Freeze plain objects
    for (const key of Object.keys(obj)) {
        deepFreeze(obj[key]); // recurse
    }
    return Object.freeze(obj);
}

/**
 * @typedef {Object} SecurityPolicy
 * @property {string[]} [allowedHostnames] - Allowed hostnames (non-IP addresses)
 * @property {string[]} [allowedIpv4Cidrs] - Allowed IPv4 CIDR ranges (e.g., "192.168.1.0/24")
 * @property {string[]} [allowedDestinationRegex] - Regex patterns for host:port strings
 */

/**
 * Convert a glob pattern to a RegExp
 * Supports: * (any chars), ? (single char), [abc] (char class), [!abc] (negated char class)
 * @param {string} glob - Glob pattern
 * @returns {RegExp} Compiled regex
 */
function globToRegex(glob) {
    let regex = '';
    let i = 0;
    
    while (i < glob.length) {
        const char = glob[i];
        
        switch (char) {
            case '*':
                // * matches any sequence of characters
                regex += '.*';
                break;
            case '?':
                // ? matches exactly one character
                regex += '.';
                break;
            case '[':
                // Character class - find closing bracket
                let j = i + 1;
                let classContent = '';
                let negated = false;
                
                // Check for negation
                if (glob[j] === '!' || glob[j] === '^') {
                    negated = true;
                    j++;
                }
                
                // Collect characters until ]
                while (j < glob.length && glob[j] !== ']') {
                    // Escape special regex chars inside class (except - and ])
                    if (glob[j] === '\\' && j + 1 < glob.length) {
                        classContent += '\\' + glob[j + 1];
                        j += 2;
                    } else {
                        classContent += glob[j];
                        j++;
                    }
                }
                
                if (j < glob.length && glob[j] === ']') {
                    regex += '[' + (negated ? '^' : '') + classContent + ']';
                    i = j;
                } else {
                    // No closing bracket - treat [ as literal
                    regex += '\\[';
                }
                break;
            case '\\':
                // Escape next character
                if (i + 1 < glob.length) {
                    regex += '\\' + glob[i + 1];
                    i++;
                } else {
                    regex += '\\\\';
                }
                break;
            // Escape regex special characters
            case '.':
            case '+':
            case '^':
            case '$':
            case '{':
            case '}':
            case '(':
            case ')':
            case '|':
                regex += '\\' + char;
                break;
            default:
                regex += char;
        }
        i++;
    }
    
    return new RegExp('^' + regex + '$', 'i');
}

/**
 * Check if a string matches a glob pattern
 * @param {string} str - String to test
 * @param {string} pattern - Glob pattern
 * @returns {boolean} True if matches
 */
function matchGlob(str, pattern) {
    try {
        const regex = globToRegex(pattern);
        return regex.test(str);
    } catch (e) {
        console.error('[RDPSecurityPolicy] Invalid glob pattern:', pattern, e);
        return false;
    }
}

/**
 * Check if a string is a valid IPv4 address
 * @param {string} str - String to check
 * @returns {boolean} True if valid IPv4
 */
function isIPv4(str) {
    const ipv4Regex = /^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$/;
    const match = str.match(ipv4Regex);
    if (!match) return false;
    
    // Validate each octet is 0-255
    for (let i = 1; i <= 4; i++) {
        const octet = parseInt(match[i], 10);
        if (octet < 0 || octet > 255) return false;
    }
    return true;
}

/**
 * Convert IPv4 address string to 32-bit integer
 * @param {string} ip - IPv4 address
 * @returns {number} 32-bit integer representation
 */
function ipv4ToInt(ip) {
    const parts = ip.split('.').map(p => parseInt(p, 10));
    return ((parts[0] << 24) | (parts[1] << 16) | (parts[2] << 8) | parts[3]) >>> 0;
}

/**
 * Check if an IPv4 address is within a CIDR range
 * @param {string} ip - IPv4 address to check
 * @param {string} cidr - CIDR notation (e.g., "192.168.1.0/24")
 * @returns {boolean} True if IP is within CIDR range
 */
function isIpInCidr(ip, cidr) {
    const [range, prefixStr] = cidr.split('/');
    const prefix = prefixStr !== undefined ? parseInt(prefixStr, 10) : 32;
    
    if (!isIPv4(range) || isNaN(prefix) || prefix < 0 || prefix > 32) {
        return false;
    }
    
    const ipInt = ipv4ToInt(ip);
    const rangeInt = ipv4ToInt(range);
    
    // Create mask from prefix length
    const mask = prefix === 0 ? 0 : (~0 << (32 - prefix)) >>> 0;
    
    return (ipInt & mask) === (rangeInt & mask);
}

/**
 * RDP Security Policy Enforcer
 * Immutable security policy that validates connection destinations
 */
export class RDPSecurityPolicy {
    /** @type {SecurityPolicy} */
    #policy;
    
    /** @type {RegExp[]} */
    #compiledRegexes;
    
    /**
     * Create a new security policy
     * @param {SecurityPolicy} policy - Security policy configuration
     * @param {boolean} [freeze=true] - Whether to deep freeze the policy
     */
    constructor(policy = {}, freeze = true) {
        // Validate and normalize policy
        const normalizedPolicy = {
            allowedHostnames: Array.isArray(policy.allowedHostnames) 
                ? [...policy.allowedHostnames] 
                : undefined,
            allowedIpv4Cidrs: Array.isArray(policy.allowedIpv4Cidrs) 
                ? [...policy.allowedIpv4Cidrs] 
                : undefined,
            allowedDestinationRegex: Array.isArray(policy.allowedDestinationRegex) 
                ? [...policy.allowedDestinationRegex] 
                : undefined,
        };
        
        // Remove undefined properties
        Object.keys(normalizedPolicy).forEach(key => {
            if (normalizedPolicy[key] === undefined) {
                delete normalizedPolicy[key];
            }
        });
        
        // Pre-compile regex patterns for performance
        this.#compiledRegexes = [];
        if (normalizedPolicy.allowedDestinationRegex) {
            for (const pattern of normalizedPolicy.allowedDestinationRegex) {
                try {
                    this.#compiledRegexes.push(new RegExp(pattern));
                } catch (e) {
                    console.error('[RDPSecurityPolicy] Invalid regex pattern:', pattern, e);
                }
            }
        }
        
        // Deep freeze if requested (default: true)
        this.#policy = freeze ? deepFreeze(normalizedPolicy) : normalizedPolicy;
        
        // Freeze the compiled regexes array too
        if (freeze) {
            Object.freeze(this.#compiledRegexes);
        }
    }
    
    /**
     * Check if policy has any rules defined
     * @returns {boolean} True if at least one rule is defined
     */
    hasRules() {
        return !!(
            this.#policy.allowedHostnames?.length ||
            this.#policy.allowedIpv4Cidrs?.length ||
            this.#policy.allowedDestinationRegex?.length
        );
    }
    
    /**
     * Validate a connection destination against the security policy
     * @param {string} host - Hostname or IP address
     * @param {number} [port=3389] - Port number
     * @returns {{allowed: boolean, reason?: string}} Validation result
     */
    validate(host, port = 3389) {
        // If no rules defined, allow all connections
        if (!this.hasRules()) {
            return { allowed: true };
        }
        
        const destination = `${host}:${port}`;
        const hostIsIPv4 = isIPv4(host);
        
        // Check destination regex patterns first (applies to both hostnames and IPs)
        if (this.#policy.allowedDestinationRegex?.length) {
            for (const regex of this.#compiledRegexes) {
                if (regex.test(destination)) {
                    return { allowed: true };
                }
            }
        }
        
        // Check hostname rules (only for non-IP hosts)
        if (!hostIsIPv4 && this.#policy.allowedHostnames?.length) {
            for (const pattern of this.#policy.allowedHostnames) {
                if (matchGlob(host, pattern)) {
                    return { allowed: true };
                }
            }
        }
        
        // Check IPv4 CIDR rules (only for IP addresses)
        if (hostIsIPv4 && this.#policy.allowedIpv4Cidrs?.length) {
            for (const cidr of this.#policy.allowedIpv4Cidrs) {
                if (isIpInCidr(host, cidr)) {
                    return { allowed: true };
                }
            }
        }
        
        // No rules matched
        return {
            allowed: false,
            reason: `Connection to ${destination} blocked by security policy`
        };
    }
    
    /**
     * Get a readonly copy of the policy (for debugging)
     * @returns {SecurityPolicy} Policy copy
     */
    getPolicy() {
        // Return the frozen policy directly (it's already immutable)
        return this.#policy;
    }
}

/**
 * Create a security policy with optional deep freeze
 * @param {SecurityPolicy} policy - Policy configuration
 * @param {boolean} [freeze=true] - Whether to deep freeze
 * @returns {RDPSecurityPolicy} Security policy instance
 */
export function createSecurityPolicy(policy, freeze = true) {
    return new RDPSecurityPolicy(policy, freeze);
}

File: backend/rdp_security.py
"""
RDP Security Policy Module
Provides connection security enforcement with immutable policy configuration
"""

import os
import re
import json
import copy
import logging
from typing import Optional, List, Dict, Any, Tuple
from dataclasses import dataclass

# Default path for security policy configuration
DEFAULT_SECURITY_POLICY_PATH = "/app/security/rdp-bridge-policy.json"

# Environment variable for overriding the default path
SECURITY_POLICY_PATH_ENV = "RDP_BRIDGE_SECURITY_POLICY_PATH"

logger = logging.getLogger(__name__)


def deep_freeze(obj: Any) -> Any:
    """
    Deep freeze an object to prevent tampering.
    In Python, we use immutable types (tuple, frozenset) and readonly patterns.
    
    Args:
        obj: Object to freeze
        
    Returns:
        Frozen (immutable) version of the object
    """
    if obj is None or isinstance(obj, (int, float, str, bool)):
        return obj
    
    if isinstance(obj, list):
        return tuple(deep_freeze(el) for el in obj)
    
    if isinstance(obj, dict):
        return MappingProxyType({k: deep_freeze(v) for k, v in obj.items()})
    
    return obj


# Import MappingProxyType for read-only dict
from types import MappingProxyType


def glob_to_regex(glob: str) -> re.Pattern:
    """
    Convert a glob pattern to a RegExp.
    Supports: * (any chars), ? (single char), [abc] (char class), [!abc] (negated char class)
    
    Args:
        glob: Glob pattern
        
    Returns:
        Compiled regex pattern
    """
    regex = ''
    i = 0
    length = len(glob)
    
    while i < length:
        char = glob[i]
        
        if char == '*':
            # * matches any sequence of characters
            regex += '.*'
        elif char == '?':
            # ? matches exactly one character
            regex += '.'
        elif char == '[':
            # Character class - find closing bracket
            j = i + 1
            class_content = ''
            negated = False
            
            # Check for negation
            if j < length and glob[j] in ('!', '^'):
                negated = True
                j += 1
            
            # Collect characters until ]
            while j < length and glob[j] != ']':
                # Escape next character if backslash
                if glob[j] == '\\' and j + 1 < length:
                    class_content += '\\' + glob[j + 1]
                    j += 2
                else:
                    class_content += glob[j]
                    j += 1
            
            if j < length and glob[j] == ']':
                regex += '[' + ('^' if negated else '') + class_content + ']'
                i = j
            else:
                # No closing bracket - treat [ as literal
                regex += '\\['
        elif char == '\\':
            # Escape next character
            if i + 1 < length:
                regex += '\\' + glob[i + 1]
                i += 1
            else:
                regex += '\\\\'
        elif char in '.+^${}()|':
            # Escape regex special characters
            regex += '\\' + char
        else:
            regex += char
        
        i += 1
    
    return re.compile('^' + regex + '$', re.IGNORECASE)


def match_glob(s: str, pattern: str) -> bool:
    """
    Check if a string matches a glob pattern.
    
    Args:
        s: String to test
        pattern: Glob pattern
        
    Returns:
        True if matches
    """
    try:
        regex = glob_to_regex(pattern)
        return bool(regex.match(s))
    except Exception as e:
        logger.error(f"[RDPSecurityPolicy] Invalid glob pattern: {pattern} - {e}")
        return False


def is_ipv4(s: str) -> bool:
    """
    Check if a string is a valid IPv4 address.
    
    Args:
        s: String to check
        
    Returns:
        True if valid IPv4
    """
    ipv4_regex = re.compile(r'^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$')
    match = ipv4_regex.match(s)
    if not match:
        return False
    
    # Validate each octet is 0-255
    for i in range(1, 5):
        octet = int(match.group(i))
        if octet < 0 or octet > 255:
            return False
    return True


def ipv4_to_int(ip: str) -> int:
    """
    Convert IPv4 address string to 32-bit integer.
    
    Args:
        ip: IPv4 address
        
    Returns:
        32-bit integer representation
    """
    parts = [int(p) for p in ip.split('.')]
    return ((parts[0] << 24) | (parts[1] << 16) | (parts[2] << 8) | parts[3]) & 0xFFFFFFFF


def is_ip_in_cidr(ip: str, cidr: str) -> bool:
    """
    Check if an IPv4 address is within a CIDR range.
    
    Args:
        ip: IPv4 address to check
        cidr: CIDR notation (e.g., "192.168.1.0/24")
        
    Returns:
        True if IP is within CIDR range
    """
    parts = cidr.split('/')
    range_ip = parts[0]
    prefix = int(parts[1]) if len(parts) > 1 else 32
    
    if not is_ipv4(range_ip) or prefix < 0 or prefix > 32:
        return False
    
    ip_int = ipv4_to_int(ip)
    range_int = ipv4_to_int(range_ip)
    
    # Create mask from prefix length
    if prefix == 0:
        mask = 0
    else:
        mask = (~0 << (32 - prefix)) & 0xFFFFFFFF
    
    return (ip_int & mask) == (range_int & mask)


@dataclass
class ValidationResult:
    """Result of a security policy validation."""
    allowed: bool
    reason: Optional[str] = None


class RDPSecurityPolicy:
    """
    RDP Security Policy Enforcer.
    Immutable security policy that validates connection destinations.
    """
    
    def __init__(self, policy: Optional[Dict[str, Any]] = None, freeze: bool = True):
        """
        Create a new security policy.
        
        Args:
            policy: Security policy configuration
            freeze: Whether to deep freeze the policy
        """
        if policy is None:
            policy = {}
        
        # Validate and normalize policy
        normalized_policy = {}
        
        if isinstance(policy.get('allowedHostnames'), list):
            normalized_policy['allowedHostnames'] = list(policy['allowedHostnames'])
        
        if isinstance(policy.get('allowedIpv4Cidrs'), list):
            normalized_policy['allowedIpv4Cidrs'] = list(policy['allowedIpv4Cidrs'])
        
        if isinstance(policy.get('allowedDestinationRegex'), list):
            normalized_policy['allowedDestinationRegex'] = list(policy['allowedDestinationRegex'])
        
        # Pre-compile regex patterns for performance
        self._compiled_regexes: List[re.Pattern] = []
        if 'allowedDestinationRegex' in normalized_policy:
            for pattern in normalized_policy['allowedDestinationRegex']:
                try:
                    self._compiled_regexes.append(re.compile(pattern))
                except Exception as e:
                    logger.error(f"[RDPSecurityPolicy] Invalid regex pattern: {pattern} - {e}")
        
        # Deep freeze if requested (default: True)
        if freeze:
            self._policy = deep_freeze(normalized_policy)
            self._compiled_regexes = tuple(self._compiled_regexes)
        else:
            self._policy = normalized_policy
    
    def has_rules(self) -> bool:
        """
        Check if policy has any rules defined.
        
        Returns:
            True if at least one rule is defined
        """
        policy = self._policy if isinstance(self._policy, dict) else dict(self._policy)
        
        allowed_hostnames = policy.get('allowedHostnames', ())
        allowed_cidrs = policy.get('allowedIpv4Cidrs', ())
        allowed_regex = policy.get('allowedDestinationRegex', ())
        
        return bool(
            (allowed_hostnames and len(allowed_hostnames) > 0) or
            (allowed_cidrs and len(allowed_cidrs) > 0) or
            (allowed_regex and len(allowed_regex) > 0)
        )
    
    def validate(self, host: str, port: int = 3389) -> ValidationResult:
        """
        Validate a connection destination against the security policy.
        
        Args:
            host: Hostname or IP address
            port: Port number (default: 3389)
            
        Returns:
            ValidationResult with allowed status and optional reason
        """
        # If no rules defined, allow all connections
        if not self.has_rules():
            return ValidationResult(allowed=True)
        
        destination = f"{host}:{port}"
        host_is_ipv4 = is_ipv4(host)
        
        policy = self._policy if isinstance(self._policy, dict) else dict(self._policy)
        
        # Check destination regex patterns first (applies to both hostnames and IPs)
        allowed_regex = policy.get('allowedDestinationRegex', ())
        if allowed_regex and len(allowed_regex) > 0:
            for regex in self._compiled_regexes:
                if regex.search(destination):
                    return ValidationResult(allowed=True)
        
        # Check hostname rules (only for non-IP hosts)
        allowed_hostnames = policy.get('allowedHostnames', ())
        if not host_is_ipv4 and allowed_hostnames and len(allowed_hostnames) > 0:
            for pattern in allowed_hostnames:
                if match_glob(host, pattern):
                    return ValidationResult(allowed=True)
        
        # Check IPv4 CIDR rules (only for IP addresses)
        allowed_cidrs = policy.get('allowedIpv4Cidrs', ())
        if host_is_ipv4 and allowed_cidrs and len(allowed_cidrs) > 0:
            for cidr in allowed_cidrs:
                if is_ip_in_cidr(host, cidr):
                    return ValidationResult(allowed=True)
        
        # No rules matched
        return ValidationResult(
            allowed=False,
            reason=f"Connection to {destination} blocked by security policy"
        )
    
    def get_policy(self) -> Dict[str, Any]:
        """
        Get a readonly copy of the policy (for debugging).
        
        Returns:
            Policy copy
        """
        # Return the frozen policy directly (it's already immutable)
        if isinstance(self._policy, MappingProxyType):
            return dict(self._policy)
        return copy.deepcopy(self._policy)


def create_security_policy(policy: Optional[Dict[str, Any]] = None, freeze: bool = True) -> RDPSecurityPolicy:
    """
    Create a security policy with optional deep freeze.
    
    Args:
        policy: Policy configuration
        freeze: Whether to deep freeze
        
    Returns:
        Security policy instance
    """
    return RDPSecurityPolicy(policy, freeze)


def load_security_policy_from_file(path: Optional[str] = None) -> RDPSecurityPolicy:
    """
    Load security policy from a JSON file.
    
    Args:
        path: Path to the policy JSON file. If None, uses environment variable
              RDP_BRIDGE_SECURITY_POLICY_PATH or defaults to /app/security/rdp-bridge-policy.json
              
    Returns:
        Security policy instance
    """
    if path is None:
        path = os.environ.get(SECURITY_POLICY_PATH_ENV, DEFAULT_SECURITY_POLICY_PATH)
    
    try:
        with open(path, 'r', encoding='utf-8') as f:
            policy_data = json.load(f)
        logger.info(f"[RDPSecurityPolicy] Loaded security policy from {path}")
        return create_security_policy(policy_data)
    except FileNotFoundError:
        logger.warning(f"[RDPSecurityPolicy] Security policy file not found: {path}, using empty policy (allow all)")
        return create_security_policy({})
    except json.JSONDecodeError as e:
        logger.error(f"[RDPSecurityPolicy] Invalid JSON in security policy file {path}: {e}")
        return create_security_policy({})
    except Exception as e:
        logger.error(f"[RDPSecurityPolicy] Error loading security policy from {path}: {e}")
        return create_security_policy({})


def get_security_policy_path() -> str:
    """
    Get the configured security policy path.
    
    Returns:
        The path from environment variable or default
    """
    return os.environ.get(SECURITY_POLICY_PATH_ENV, DEFAULT_SECURITY_POLICY_PATH)


# Singleton instance for global use
_global_security_policy: Optional[RDPSecurityPolicy] = None


def get_security_policy() -> RDPSecurityPolicy:
    """
    Get the global security policy instance, loading from file if not already loaded.
    
    Returns:
        The global security policy instance
    """
    global _global_security_policy
    if _global_security_policy is None:
        _global_security_policy = load_security_policy_from_file()
    return _global_security_policy


def reload_security_policy() -> RDPSecurityPolicy:
    """
    Reload the global security policy from file.
    
    Returns:
        The newly loaded security policy instance
    """
    global _global_security_policy
    _global_security_policy = load_security_policy_from_file()
    return _global_security_policy
